src/snowflake/connector/config_manager.py

```diff
@@ -0,0 +1,333 @@
+#
+# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+#
+
+from __future__ import annotations
+
+import logging
+import os
+import stat
+from collections.abc import Iterable
+from operator import methodcaller
+from pathlib import Path
+from typing import Any, Callable, TypeVar
+from warnings import warn
+
+import tomlkit
+from tomlkit.items import Table
+from typing_extensions import Literal
+
+from snowflake.connector.constants import CONFIG_FILE
+from snowflake.connector.errors import ConfigManagerError, ConfigSourceError
+
+_T = TypeVar("_T")
+
+LOGGER = logging.getLogger(__name__)
+READABLE_BY_OTHERS = stat.S_IRGRP | stat.S_IROTH
```

Comment:
Do we care if the file is o+wx?  I think ssh wants its directory and config file to be only u+rw but I'm not positive about that.

Reply:
I had the same dilemma... I decided to not worry about these really unusual cases, but I'm still not a 100% sure whether we should

Reply:
I would approach it from an allow list perspective then a deny list - can we just come up with a permission that _is okay_ and raise a warning for everything else. That might be simpler.

Reply:
Not sure that will work, but let's at least agree on the permissions that will be allowed or denied.

* u+rw -- this is allowed.  we probably don't care about u+x, though for strictness, probably denied if u+x (check w/ssh?)
* g+rw -- should we allow this?  I'd opt for "no" for now and we can always relax it later.  g+x same treatment as above.
* o+rwx - deny